rust/private/expand.bzl

```diff
@@ -0,0 +1,247 @@
+        crate_info = crate_info,
+        dep_info = dep_info,
+        linkstamp_outs = linkstamp_outs,
+        ambiguous_libs = ambiguous_libs,
+        output_hash = determine_output_hash(crate_info.root, ctx.label),
+        rust_flags = [],
+        out_dir = out_dir,
+        build_env_files = build_env_files,
+        build_flags_files = build_flags_files,
+        emit = ["dep-info", "metadata"],
+    )
+
+    if crate_info.is_test:
+        args.rustc_flags.add("--test")
+
+    expand_out = ctx.actions.declare_file(ctx.label.name + ".expand.rs", sibling = crate_info.output)
+    args.process_wrapper_flags.add("--stdout-file", expand_out.path)
+
+    # Expand all macros and dump the source to stdout.
+    args.rustc_flags.add("-Zunpretty=expanded")
+
+    ctx.actions.run(
+        executable = ctx.executable._process_wrapper,
+        inputs = compile_inputs,
+        outputs = [expand_out],
+        env = env,
+        arguments = args.all,
+        mnemonic = "RustExpand",
+    )
+
+    return [
+        OutputGroupInfo(expanded = depset([expand_out])),
+    ]
+
+# Example: Expand all rust targets in the codebase.
+#   bazel build --aspects=@rules_rust//rust:defs.bzl%rust_expand_aspect \
+#               --output_groups=expanded \
+#               //...
+rust_expand_aspect = aspect(
+    fragments = ["cpp"],
+    host_fragments = ["cpp"],
+    attrs = {
+        "_cc_toolchain": attr.label(
+            doc = (
+                "Required attribute to access the cc_toolchain. See [Accessing the C++ toolchain]" +
+                "(https://docs.bazel.build/versions/master/integrating-with-rules-cc.html#accessing-the-c-toolchain)"
+            ),
+            default = Label("@bazel_tools//tools/cpp:current_cc_toolchain"),
+        ),
+        "_extra_rustc_flag": attr.label(default = "//:extra_rustc_flag"),
+        "_extra_rustc_flags": attr.label(default = "//:extra_rustc_flags"),
+        "_process_wrapper": attr.label(
+            doc = "A process wrapper for running clippy on all platforms",
+            default = Label("//util/process_wrapper"),
+            executable = True,
+            cfg = "exec",
+        ),
+    },
+    toolchains = [
+        str(Label("//rust:toolchain_type")),
+        "@bazel_tools//tools/cpp:toolchain_type",
+    ],
+    incompatible_use_toolchain_transition = True,
+    implementation = _expand_aspect_impl,
+    doc = """\
+Executes Rust expand on specified targets.
+
+This aspect applies to existing rust_library, rust_test, and rust_binary rules.
+
+As an example, if the following is defined in `examples/hello_lib/BUILD.bazel`:
+
+```python
+load("@rules_rust//rust:defs.bzl", "rust_library", "rust_test")
+
+rust_library(
+    name = "hello_lib",
+    srcs = ["src/lib.rs"],
+)
+
+rust_test(
+    name = "greeting_test",
+    srcs = ["tests/greeting.rs"],
+    deps = [":hello_lib"],
+)
+```
+
+Then the targets can be expanded with the following command:
```

Comment:
could you add a small snippet of how someone might inspect the generated source after running this?